### ISSUE TITLE
Store: consumes lots of memory at startup and loop restart cause OOM

### ISSUE THREAD

**Issue by chalut01** [thanos-io/thanos#6643]
Thanos Version v0.32.0
Everything is set up the same way, only the version is different. Are there others similar to me?

![Screenshot 2566-08-24 at 13 33 49](https://github.com/thanos-io/thanos/assets/22786179/0aeeb355-9ae2-4cc4-9b2a-0328c2bd5294)

**Comment by rgarcia89** [comment]
I have experienced exactly the same issue.

Closed my ticket as this one covers the same issue https://github.com/thanos-io/thanos/issues/6644

**Comment by pahaeanx** [comment]
Same here. Can't get it to start up with 6GB of RAM when it used to run with ~3GB.

**Comment by yeya24** [comment]
I can only think about https://github.com/thanos-io/thanos/pull/6509 that might change the startup memory usage of store gateway. But ideally this change should improve the mem usage.

Do you enable lazy index header? Can you share the config of store gateway?

**Comment by pahaeanx** [comment]
> I can only think about #6509 that might change the startup memory usage of store gateway. But ideally this change should improve the mem usage.
> 
> Do you enable lazy index header? Can you share the config of store gateway?

No, I use a pretty vanilla config I'd say. Failing config (formatted for readability):

```
thanos store --max-time=-1w --grpc-address=localhost:15000 --http-address=localhost:15001 \ 
--data-dir=/var/lib/thanos-cache/store  --objstore.config-file=/etc/thanos/s3.yml \
--grpc-server-tls-cert=/etc/thanos/thanos.cer --grpc-server-tls-key=/etc/thanos/thanos.key \
--sync-block-duration=30m
```
This is currently running with `0.30.2` and fails with `0.32.0`.

**Comment by rgarcia89** [comment]
@yeya24 I deploy the thanos store using kube-thanos. It starts with the following args in the manifest

```
        - store
        - --log.level=info
        - --log.format=logfmt
        - --data-dir=/var/thanos/store
        - --grpc-address=0.0.0.0:10901
        - --http-address=0.0.0.0:10902
        - --objstore.config=$(OBJSTORE_CONFIG)
        - --ignore-deletion-marks-delay=24h
 ```

**Comment by antikilahdjs** [comment]
I have the same issue and upgraded to v31, all the other components works perfectly in 0.32 version but the store consume 1tb

**Comment by GiedriusS** [comment]
Maybe it would be possible for you to take a pprof memory profile during bootup of Thanos Store and share it here? Thanos Store exposes a pprof endpoint on `/debug/pprof`. :thinking:

**Comment by bboysoulcn** [comment]
the same as you

**Comment by chalut01** [comment]
@GiedriusS  I'm sorry, but I can't actually try out andget pprof on production environment. 
However, In my development environment, It have a small data and everything works well with v0.32.0.

Can someone share a pprof?

**Comment by rgarcia89** [comment]
> Maybe it would be possible for you to take a pprof memory profile during bootup of Thanos Store and share it here? Thanos Store exposes a pprof endpoint on `/debug/pprof`. 🤔

@GiedriusS I have thanos running as container in kubernetes using kube-thanos. How can I find this endpoint?

**Comment by yeya24** [comment]
@rgarcia89 Can you port forward one of your store gateway pod using its http port, for example it is 8080.

```
curl http://localhost:8080/debug/pprof/heap > heap.pprof
```

You can get the heap profile by running the command above.

**Comment by rgarcia89** [comment]
@yeya24 here you go - including the heap of running 0.31.0 thanos store vs the heap of a just started 0.32.0 thanos store which then crashed

[heap-pprof.zip](https://github.com/thanos-io/thanos/files/12451747/heap-pprof.zip)

**Comment by MichaHoffmann** [comment]
Oh thats the cuckoo filter completely.

**Comment by MichaHoffmann** [comment]
Do you have some figures around your cardinality?

**Comment by rgarcia89** [comment]
@MichaHoffmann anything specific you are looking for? Thanos is currently managing a time series of 1792044 entries

**Comment by MichaHoffmann** [comment]
This filter should scale ( theoretically ) with the amount of label names you have. But for 128gb it would need like 100s of millions ( again theoretically, maybe there is a bug somewhere )

**Comment by rgarcia89** [comment]
Good question. I can see everything running smooth with v0.31.0 and also haven't seen any kind of alerts or issues in regards of high cardinality :man_shrugging:

**Comment by MichaHoffmann** [comment]
> Good question. I can see everything running smooth with v0.31.0 and also haven't seen any kind of alerts or issues in regards of high cardinality

That cuckoo filter was introduced in 0.32.0 so that makes sense

**Comment by rgarcia89** [comment]
Maybe there is something counting wrong in that filter. Otherwise I am not sure how 128gb can be justified

**Comment by saswatamcode** [comment]
https://github.com/thanos-io/thanos/pull/6669 should address this

**Comment by pahaeanx** [comment]
Unfortunately still crashing with `0.32.1`. Can't do any more digging today so all I can offer for now is the crash log. But it seems like it OOM crashed again. I can try and scale the VM tomorrow so I can maybe get it to start up.


[crash.log.tar.gz](https://github.com/thanos-io/thanos/files/12455023/crash.log.tar.gz)

**Comment by mateuszdrab** [comment]
Works for me now, I can check memory metrics later.

**Comment by antikilahdjs** [comment]
Works perfectly and thank you guys to work hard to delivery the best to us

**Comment by MichaHoffmann** [comment]
@chalut01 can we close the issue?

**Comment by chalut01** [comment]
Fixed in v0.32.2 !!
Thank you everyone for your hard work.
